Add scroll_container formspec element (redo)

Desour opened the pull request:
Does what is described in #8576 and closes #8576.
This is a redo of #8591 (rebasing would probably have been more difficult). Many things are copied from that PR.
Note that #8740 has made this PR much easier.

- There's the new class `GUIScrollContainer`.
- The scroll_container consists of two elements. One clips every child, the other moves them. The clipper is the parent of the mover.
- The scroll container mover (of class `GUIScrollContainer`) is moved on event.

## To do

This PR is a Ready for Review.

## How to test

Use `/formspec` in minimal.
old:
<details><summary>Here's a formspec that I've used for testing.</summary>

```lua
local my_formspec =
	"formspec_version[3]"..
	"size[10,8]"..
	--~ "real_coordinates[true]"..
	--~ "position[0.2,0.7]"..
	--~ "anchor[0,1]"..
	"button[8,1;1,1;bla;Bla]"..
	--~ "scrollbaroptions[max=100]"..
	"box[1,1;8,6;#00aa]"..
	"scroll_container[1,1;8,6;scrbar;down]"..
		"button[0,1;1,1;baum;Baum]"..
		--~ "container[0,0]"..
		"button[0,10;1,1;planze;Pflanze]"..
		"pwdfield[2,2;1,1;baum2;Baum]"..
		--~ "container_end[]"..
		"list[current_player;main;4,4;1,5;]"..
		"box[2,5;3,2;#ffff00]"..
		"image[1,10;3,2;bubble.png]"..
		"image[3,1;bubble.png]"..
		"item_image[2,6;3,2;default:mese]"..
		"label[2,15;bla baum\nfoo bar]"..
		--~ "textarea[2,10;5,6;txtarea;Baum;bla]"..
		"item_image_button[2,3;1,1;default:dirt_with_grass;itemimagebutton;ItemImageButton]"..
		--~ "tooltip[planze;Blatt;#f00;#000]"..
		"tooltip[0,11;3,2;Blatt;#f00;#000]"..
		"box[0,11;3,2;#00ff00]"..
		--~ "checkbox[2,25;chb;ChB;false]",
		"container[0,18]"..
			"box[1,2;3,2;#0a0a]"..
			"scroll_container[1,2;3,2;scrbar2;right;0.06]"..
				"label[10,0;nest]"..
				"button[0,0;6,1;butnest;Nest]"..
				--~ "bgcolor[#aa0a]"..
			"scroll_container_end[]"..
			"scrollbar[1,0;3.5,0.3;horizontal;scrbar2;0]"..
		"container_end[]"..
		--~ "tablecolumns[text,bla]"..
		--~ "table[1,23;3,1;tbl;a,b,c,d;2]"..
		--~ "textlist[0,5;2,0.5;txtlst;a,b,c]"..
		"dropdown[0,6;2;hmdrpdwn;apfel,birne;1]"..
		"image_button[0,4;2,2;bubble.png;bubblebutton;bbbbtt;false;true;heart.png]"..
		--~ "tabheader[-1,9;mytabheader;a,b,c;2;false;false]"..
		--~ "background[2,4;2,2;default_wood.png;true]"..
		--~ "bgcolor[#00aa]"..
		"box[1,22.5;4,1;#a00a]"..
	"scroll_container_end[]"..
	"scrollbaroptions[max=170]".. -- lowest seen pos is: 0.1*170+6=23 (factor*max+height)
	"scrollbar[7.5,0;0.3,4;vertical;scrbar;0]"..
	"scrollbar[8,0;0.3,4;vertical;scrbarhmmm;0]"..
	"dropdown[0,6;2;hmdrpdwnnn;apfel,birne;1]"..
	--~ "background[5,5;1,1;testbg.png;true;10]"..
	--~ "textarea[5,0.5;1,1;scrtxtarea;Baum;blablaajk adfkjgn \n akfjb]"..
	--~ "bgcolor[;true]"..
	--~ "bgcolor[#0000]"..
	""
```
</details>

Here's an (old) demo video: [scroll_container_show.m4v.zip](https://github.com/minetest/minetest/files/3273388/scroll_container_show.m4v.zip)

DS-Minetest commented:
Rebased.

Hm, currently I'm unsure what to do with `bgcolor`. The old PR changed it so that a box is added as background to the `scroll_container` if `bgcolor` is used in a `scroll_container`, but now I don't know what benefits this would have over just adding a box element before the `scroll_container` in the formspec.

v-rob commented:
`bgcolor` useless to add since boxes have proper stacking, so I'm in favor of less code for a useless thing.

rubenwardy commented:
I'd avoid adding unessential features like that

This PR is next on the formspec review list, would appreciate an update :)

DS-Minetest commented:
Ok, now it should be rebased correctly.
(Btw. github doesn't make notifications when labels are added.)

DS-Minetest commented:
Would someone like to remove the `WIP` and `Rebase needed` labels btw? They don't apply anymore.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -2037,6 +2037,24 @@ Elements
 
+### `scroll_container[<X>,<Y>;<W>,<H>;<scrollbar name>;<direction>;<scroll factor>]`
+
+* Start of a scroll_container block, all elements in a scroll_container block will
```
```suggestion
* Start of a scroll_container block. All contained elements will ...
```

DS-Minetest replied:
Done.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -2037,6 +2037,24 @@ Elements
+### `scroll_container[<X>,<Y>;<W>,<H>;<scrollbar name>;<direction>;<scroll factor>]`
+
+* Start of a scroll_container block, all elements in a scroll_container block will
+  * take the block's coordinate as position origin,
```
```suggestion
  * take the scroll container coordinate as position origin
```

DS-Minetest replied:
Why should the comma be removed?

SmallJoker replied:
No particular reason. Probably happened by re-writing the line.

DS-Minetest replied:
I'll mark this as resolved.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -347,6 +348,99 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+			container_stack.empty()) {
+		errorstream << "Invalid scroll_container end element, "
+				<< "no matching scroll_container start element" << std::endl;
+	} else if (pos_offset.getLengthSQ() != 0.0f) {
```
When does this happen?
What if the scroll container is at `0,0` ?

DS-Minetest replied:
`pos_offset` is only used by normal containers. `scroll_container`s set the `pos_offset` to (0,0) so that their children aren't affected by this offset. If the offset length is not 0 here, there was an unclosed normal container before.
I'll add a code comment.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -347,6 +348,99 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+	if (data->current_parent == this || data->current_parent->getParent() == this ||
+			container_stack.empty()) {
+		errorstream << "Invalid scroll_container end element, "
+				<< "no matching scroll_container start element" << std::endl;
```
Please add an early return here and drop `else` below to clarify the execution structure. Nothing may happen if there's no valid container to close.

DS-Minetest replied:
Done.

SmallJoker commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -738,10 +832,12 @@ void GUIFormSpecMenu::parseImage(parserData* data, const std::string &element)
 		e->setScaleImage(true);
-		e->setNotClipped(true);
+		if (data->current_parent == this)
+			e->setNotClipped(true);
```
How about styles? Probably needs rebase?

DS-Minetest replied:
I've rebased, there were no conflicts (and I couldn't find any new place where `this` used as parent).
#9262 will probably cause a conflict here.
:thinking: This `if` here was probably a bad idea. When #9262 is merged, the not clipping will only be the default for old formspec versions (which don't support `scroll_container`s anyway). This will make the change here unneeded, so I'll remove it.

SmallJoker commented on src/gui/guiScrollContainer.h:
```diff
@@ -0,0 +1,49 @@
+private:
+	GUIScrollBar *m_scrollbar;
+	std::string m_direction;
+	f32 m_scrollfactor;
```
You could replace both `m_scrollfactor` and `m_direction` with `v2f32` which points either down or right.

DS-Minetest replied:
But then I would have to check the components of that vector. I can't simply add a vector to `rect.UpperLeftCorner` as the scroll position is given absolute.

I've replaced the string with an enum here.

SmallJoker replied:
Alright, I hoped it would be simpler than that solution. Good enough I guess :man_shrugging:

DS-Minetest commented:
>Additionally: similar to textarea[], scroll_container[] should best add the scroll bar automatically. Manual placement is annoying and forgetting it upsets Android players (no scroll wheel).

- `scroll_container`s don't work without a scrollbar, so there's no reason to worry about android players.
- Automatically added scrollbars would give less freedom.
- Now that we have real coordinates, placement should be easy.

DS-Minetest commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2091,7 +2088,7 @@ void GUIFormSpecMenu::parseTabHeader(parserData* data, const std::string &elemen
 
 		auto style = getStyleForElement("tabheader", name);
-		e->setNotClipped(style.getBool(StyleSpec::NOCLIP, data->current_parent == this));
+		e->setNotClipped(style.getBool(StyleSpec::NOCLIP, true));
```
I someone really wants to use a tabheader element in a `scroll_container`, they should step it clipped themselves.

DS-Minetest commented:
Thanks for reviewing!
Rebased and fixed comments as good as I could.

DS-Minetest commented:
Rebased again.

ClobberXD commented:
Nice work! Upon testing (with the formspec provided in the first post) I noticed a couple of things:

- Some formspec elements like buttons, lists, and dropdowns are clipped while images and boxes aren't:
  ![screenshot_20200117_151128](https://user-images.githubusercontent.com/36130650/72601700-ee2cf980-393b-11ea-8a0f-9c162ce8e507.png)
  ![screenshot_20200117_151120](https://user-images.githubusercontent.com/36130650/72601701-ee2cf980-393b-11ea-97f4-ce119feeb404.png)
  ![screenshot_20200117_151112](https://user-images.githubusercontent.com/36130650/72601702-eec59000-393b-11ea-964e-f6e96815bebb.png)
- Both the horizontal and vertical scrollbars allow scrolling _way_ beyond the actual content.

I'm not sure if either of these are unintended or even related to this PR, but I just wanted to seek clarification regarding these.

Df458 commented:
> Some formspec elements like buttons, lists, and dropdowns are clipped while images and boxes aren't

I'm not the author, but I believe that is to be expected. This is an issue that applies to those elements normally, one of my PRs (#9262) just fixed this and I expect that the code isn't rebased on that yet. It would be interesting to cherry-pick the fix locally for testing, however.

ClobberXD commented:
> > Some formspec elements like buttons, lists, and dropdowns are clipped while images and boxes aren't
> 
> I'm not the author, but I believe that is to be expected. This is an issue that applies to those elements normally, one of my PRs (#9262) just fixed this and I expect that the code isn't rebased on that yet. It would be interesting to cherry-pick the fix locally for testing, however.

Ah, nice. Thanks for clarifying.

DS-Minetest commented:
Thanks for testing, @ClobberXD!

> Some formspec elements like buttons, lists, and dropdowns are clipped while images and boxes aren't:

You have to use `formspec_version[3]` (as @Df458 already noted) or enable clipping for these elements with style.
I'll update the test code in the first post.

> Both the horizontal and vertical scrollbars allow scrolling _way_ beyond the actual content.

That's the expected behaviour. `scrollbaroptions` should be used here. I'll add a note in the doc.

ClobberXD commented:
> You have to use formspec_version[3] (as @Df458 already noted) or enable clipping for these elements with style.

Yes, I'll test this again after tweaking the test script and let you know.

> That's the expected behaviour. scrollbaroptions should be used here. I'll add a note in the doc.

Righty-ho, good to know. Thanks!

ClobberXD commented:
Cherry-picked 680aaa1 into this branch. The boxes and images clip properly now, but I notice that an image of a meseblock is stuck to the bottom of the formspec, and doesn't scroll with the rest of the container. Also, only the very top of the vertical scrollbar is clickable, and clicking any further down the scrollbar does not allow me to interact with it.

![screenshot_20200118_083822](https://user-images.githubusercontent.com/36130650/72657767-617f4b80-39ce-11ea-84df-d072130ab1e3.png)

I only added `formspec_version[3]` to the start of the formspec string, and didn't modify anything else in the test formspec from the first post. It's to be noted that there were a couple of merge conflicts when cherry-picking the aforementioned commit, and I probably messed up some of the code while resolving the conflicts (I don't understand much of the GUI code). Do let me know if I'm mistaken.

DS-Minetest commented:
> Cherry-picked [680aaa1](https://github.com/minetest/minetest/commit/680aaa1bd55fd4f0162ba31ca47d7e5b3563f2a1) into this branch.

That shouldn't be needed, I've rebased after that PR was merged.

> The boxes and images clip properly now, but I notice that an image of a meseblock is stuck to the bottom of the formspec, and doesn't scroll with the rest of the container.

Can't reproduce.

> Also, only the very top of the vertical scrollbar is clickable, and clicking any further down the scrollbar does not allow me to interact with it.

That's because in the test formspec the scrollbar was behind the `scroll_container`, sorry.
I've updated the test formspec. (There's also a max now.)

ClobberXD commented:
> That shouldn't be needed, I've rebased after that PR was merged.

Oh lol. That explains the oddities. Will retest again. :)

ClobberXD commented:
Merge conflict. I think I'm the reason for this. Should be pretty trivial to resolve though.

DS-Minetest commented:
Rebased.

rubenwardy commented:
Rebased here: https://github.com/rubenwardy/minetest/tree/scroll_container2

rubenwardy commented on src/gui/guiScrollContainer.cpp:
```diff
@@ -0,0 +1,58 @@
+/*
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
```
wrong copyright

DS-Minetest replied:
I'm not good at such things. Feel free to suggest something.

rubenwardy replied:
>  Copyright (C) 2020 DS-Minetest

rubenwardy commented on src/gui/guiScrollContainer.h:
```diff
@@ -0,0 +1,56 @@
+/*
+Minetest
+Copyright (C) 2013 celeron55, Perttu Ahola <celeron55@gmail.com>
```
wrong copyright

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -347,6 +348,111 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+		// not 0,0, it is a normal container that was opened last, not a
+		// scroll_container
+		errorstream << "Invalid scroll_container end element, "
+				<< "try to close normal container start element" << std::endl;
```
an inner container was left open

DS-Minetest replied:
Done.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -347,6 +348,111 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+
+void GUIFormSpecMenu::parseScrollContainerEnd(parserData *data)
+{
+	if (data->current_parent == this || data->current_parent->getParent() == this ||
```
I'm not a fan of how this works, I think it would be better to make the container_stack take a struct rather than positions, which contains meta information on the container types. However, I'm fine with this given there's only one special container type currently

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -347,6 +348,111 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+	// ^ is parent of other elements
+
+	// make clipper
+	FieldSpec spec_clipper(
```
Why do you need to allocate an ID for the clipper and mover?

DS-Minetest replied:
Afaik, this is done for every element.
What id should I use instead for the clipper?

rubenwardy replied:
I admit that the scope of fieldspecs confuse me. To clarify - is there a reason why the id isn't -1 with no fieldspec?

DS-Minetest replied:
Hm, I don't think, anything would speak against -1.
Should I use -1?

SmallJoker replied:
`FieldSpec` entries are only needed if the element....

* has a name (for styling, sending fields or tooltips)
* or needs special legacy sorting
* ~or has clipped text/label that is shown entirely using a tooltip~
   * ^ possible use-case in the future, not implemented.

All other `m_fields` entries are superfluous because they neither contain information nor provide any special functionality.

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -971,22 +1079,34 @@ void GUIFormSpecMenu::parseBackground(parserData* data, const std::string &eleme
 			// no auto_clip => position like normal image
-			rect = core::rect<s32>(pos, pos + geom);
-		} else {
+			core::rect<s32> rect(pos, pos + geom);
```
Please change this to use the old design, where you set variables then only create the GUI element in one place. This makes it *a lot* easier to read and reason about. You can set the clip variable in these if-statements

DS-Minetest replied:
Done.

rubenwardy commented:
Please also add a new tab to the formspec shown using `/formspec` and defined in `games/minimal/test/formspec.lua`

DS-Minetest commented:
Rebased and addressed most of the comments.
I'm sorry that I don't really have time for minetest at the moment.

rubenwardy commented on src/gui/guiScrollContainer.h:
```diff
@@ -0,0 +1,56 @@
+private:
+	enum DirectionEnum
+	{
+		DOWN,
```
please use `VERTICAL` and `HORIZONTAL` to be consistent with the scrollbar element

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2071,6 +2071,27 @@ Elements
 * End of a container, following elements are no longer relative to this
   container.
 
+### `scroll_container[<X>,<Y>;<W>,<H>;<scrollbar name>;<direction>;<scroll factor>]`
```
call it `<orientation>`

rubenwardy commented on doc/lua_api.txt:
```diff
@@ -2071,6 +2071,27 @@ Elements
+  * be additionally moved by the current value of the scrollbar with the name
+    `scrollbar name` times `scroll factor` in the direction `direction` and
+  * be clipped to the rectangle defined by `X`, `Y`, `W` and `H`.
+* Possible values for `direction` are `down` and `right`.
```
```suggestion
* Possible values for `orientation` are `vertical` and `horizontal`.
```

rubenwardy commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -349,6 +350,111 @@ void GUIFormSpecMenu::parseContainerEnd(parserData* data)
+
+	if (pos_offset.getLengthSQ() != 0.0f) {
+		// pos_offset is only set by containers and scroll_containers.
+		// scroll_containers allways set it to 0,0 which means that if it is
```
```suggestion
		// scroll_containers always set it to 0,0 which means that if it is
```

DS-Minetest commented:
I've renamed the directions, as requested. (I guess it is clear, that it scrolls down respectively to the right.)

rubenwardy commented:
Bug: if you start scrolling whilst an item slot is hovered, the tooltip will stay until it is replaced by another tooltip

DS-Minetest commented:
You're right. Buttons also stay hovered after scrolling.
I'll try to fix this.

DS-Minetest commented:
I've tried saving the last mouse move event in `GUIFormspecMenu` (`m_last_mouse_move_event`) and re-sending it in (the end of) `GUIFormspecMenu::OnEvent` like this:
```cpp
if (scroll_container_was_moved_in_this_OnEvent) {
		Environment->postEventFromUser(m_last_mouse_move_event); // needed for buttons (env->getHovered()) to be updated
		Environment->getRootGUIElement()
				->getElementFromPoint(
						core::position2d<s32>(m_last_mouse_move_event.MouseInput.X,
								m_last_mouse_move_event.MouseInput.Y))
				->OnEvent(m_last_mouse_move_event); // needed for GUIInventorLists to update their hovered_i (postEventFromUser doesn't seem to call OnEvent)
}
```
But then scrolling using the scrollbar slider didn't work anymore.

Hence I don't know how to fix this properly.
Imo the bug is ugly, but acceptable, it doesn't really hurt.

Desour commented:
:/ `background[]`s to scroll containers don't seem to work nicely. I'll simply remove the changes to `background[]` as an `image[]` element is much better here and ordering elements is possible now.

Desour commented:
Everything should be fine now.
Apart from the comment that -1 could be used as ID for the clipper, I can't decide here.

kilbith commented:
Question: what happens if a `scroll_container` is inside another one?

rubenwardy commented:
I can confirm that your changes fix the named and area tooltips, but not the item tooltips.

To fix this cleanly, we could do with a refactor of how tooltips are set. It should be done using one method - GUIFormspecMenu should iterate through elements under the cursor, calling a `getTooltip()` method. The first one which returns a non-empty string wins

Desour commented:
I didn't push any changes regarding the tooltip bug.

The area and named tooltips are implemented in GUIFormspecMenu by checking the mouse pos on every draw. The tooltip in InventoryLists (= the index of the selected item) and whether a button is hovered is determined when the mouse moves (there are also events (guievents) for when the cursor leaves or enters the rect of an element).

Firing a mouse event was what I was trying to do, but as written before that resulted in problems (scrolling by pulling the scrollbar slider somehow didn't work anymore).
(Updating where the mouse is relative to other elements should also be done because of other reasons, eg. hovered and pressed buttons stay hovered respectively pressed.)

Calling something like `getTooltip()` on every draw from the elements at the mousepos (in a particular order) is indeed a good idea. There's also a [member function in IGUIElement](http://irrlicht.sourceforge.net/docu/classirr_1_1gui_1_1_i_g_u_i_element.html#aaaac5a17bb0f0d7dcdc8709a800a6ae0) (with wide strings). Lists currently do it the other way around: https://github.com/minetest/minetest/blob/0e88176db8ce9f3fb631feb56bdd29d4ada054f5/src/gui/guiInventoryList.cpp#L147
However, I don't think that belongs into this issue.

rubenwardy commented:
This is gross, but works:

```cpp.diff
diff --git a/src/gui/guiScrollContainer.cpp b/src/gui/guiScrollContainer.cpp
index f58161e1a..cef64588d 100644
--- a/src/gui/guiScrollContainer.cpp
+++ b/src/gui/guiScrollContainer.cpp
@@ -36,9 +36,30 @@ GUIScrollContainer::GUIScrollContainer(gui::IGUIEnvironment *env,
 bool GUIScrollContainer::OnEvent(const SEvent &event)
 {
        if (event.EventType == EET_MOUSE_INPUT_EVENT &&
+                       event.MouseInput.Event == EMIE_MOUSE_MOVED) {
+               lastMousePosition.X = event.MouseInput.X;
+               lastMousePosition.Y = event.MouseInput.Y;
+       } else if (event.EventType == EET_MOUSE_INPUT_EVENT &&
                        event.MouseInput.Event == EMIE_MOUSE_WHEEL && m_scrollbar) {
                Environment->setFocus(m_scrollbar);
-               return m_scrollbar->OnEvent(event);
+
+               if (!m_scrollbar->OnEvent(event)) {
+                       return false;
+               }
+
+               // HACK: this is gross, but is needed to allow children to update hovers
+               irr::SEvent evt{};
+               evt.EventType = EET_MOUSE_INPUT_EVENT;
+               evt.MouseInput = irr::SEvent::SMouseInput{};
+               evt.MouseInput.Event = irr::EMOUSE_INPUT_EVENT::EMIE_MOUSE_MOVED;
+               evt.MouseInput.X = lastMousePosition.X;
+               evt.MouseInput.Y = lastMousePosition.Y;
+               for (auto child : this->getChildren()) {
+                       if (child->OnEvent(evt)) {
+                               return true;
+                       }
+               }
+               return true;
        }
 
        return IGUIElement::OnEvent(event);
diff --git a/src/gui/guiScrollContainer.h b/src/gui/guiScrollContainer.h
index 835a661bf..e50621a8c 100644
--- a/src/gui/guiScrollContainer.h
+++ b/src/gui/guiScrollContainer.h
@@ -51,6 +51,7 @@ class GUIScrollContainer : public gui::IGUIElement
        GUIScrollBar *m_scrollbar;
        OrientationEnum m_orientation;
        f32 m_scrollfactor;
+       irr::core::vector2di lastMousePosition {};
 
        void updateScrolling();
 };
```

rubenwardy commented:
> there are also events (guievents) for when the cursor leaves or enters the rect of an element

Would it be possible to change inventory slots to use this system if inventory slots were elements in their own right?

Desour commented:
Fixed the bug mostly.
The only problem that still seems to persists is that you can press a button, scroll and it will stay in place. I guess this is because the focus gets changed.

A bit of irc log here: http://irc.minetest.net/minetest-dev/2020-03-08#i_5649322

Desour commented:
>The only problem that still seems to persists is that you can press a button, scroll and it will stay in place.

Fixed.

Desour commented:
Squashed and rebased.

(And also changed copyright name back to "DS" and added drop calls to `m_scroll_containers` in the GUIFormspecMenu destructor (I wonder why it was missing).)

Desour commented:
Would someone mind removing the `Rebase needed` label?

rubenwardy commented:
Sorry lol!

Desour commented:
Rebased again.

rubenwardy commented:
Code looks good to me.

Hypertext elements aren't clipped. There's also weird cursor behaviour when hovering over an anchor outside the scroll container - it flickers between clickable and not

![image](https://user-images.githubusercontent.com/2122943/79056018-f69a8500-7c49-11ea-9641-59968feffe44.png)

```
diff --git a/games/minimal/mods/test/formspec.lua b/games/minimal/mods/test/formspec.lua
index b96c5b13d..2059e6f33 100644
--- a/games/minimal/mods/test/formspec.lua
+++ b/games/minimal/mods/test/formspec.lua
@@ -1,5 +1,34 @@
 local color = minetest.colorize
 
+local hypertext = minetest.formspec_escape([[^M
+       <global halign=justify valign=center background=#CCF color=#444 actioncolor=darkblue margin=10>^M
+<center><bigger>Furnace</bigger></center>^M
+<style color=black>Furnaces</style> are <action name="crafting">crafted</action> and used by the player for the purpose of cooking food and <action name="smelting">smelting</action> various items.^M
+<item name=default:furnace float=right width=128 rotate=yes>^M
+<style color=black>Type:</style> Solid block^M
+<style color=black>Drops:</style> Itself^M
+<style color=black>Physics:</style> No^M
+<style color=black>Luminance:</style> Inactive:No Active:Yes (8)^M
+<style color=black>Flammable:</style> No^M
+<style color=black>Generated:</style> No^M
+<style color=black>Renewable:</style> Yes^M
+<style color=black>Stackable:</style> Yes (99)^M
+<style color=black>Itemstring:</style> default:furnace default:furnace_active^M
+<big>Usage</big>^M
+The furnace menu can be accessed by <action name="using">using</action> the furnace.^M
+The furnace has 3 <action name="inventory">inventories</action>: An input slot, a fuel slot and 4 output slots. The fire in the furnace will automatically start when there is a smeltable item in the input slot and a fuel in the fuel slot.^M
+As long as the fire is on, the furnace will smelt any smeltable item in the input slot, one by one, until it is empty. When the fire goes off, it will smelt the next item until there are no smeltable items and no fuel items left.^M
+The current stage of cooking can be seen by <action name="pointing">pointing</action> the furnace or by viewing the furnace menu. In the furnace menu, the flame symbol roughly shows the remaining burning time. The arrow symbol shows the progress of the current smelting process.^M
+<big>Renewing</big>^M
+Furnaces can be crafted from e.g. <action name="default:cobblestone">cobblestone</action>, a renewable resource.^M
+<big>Crafting</big>^M
+Sorry no way to display crafting yet in formspec pages.^M
+<big>Fuel</big>^M
+See <action name="smelting">Smelting</action> for a list of furnace fuels.^M
+<big>Recipes</big>^M
+See the <action name="smelting">Smelting</action> page.^M
+]])^M
+^M
 local clip_fs = [[
        style_type[label,button,image_button,item_image_button,
                        tabheader,scrollbar,table,animated_image
@@ -207,6 +236,7 @@ Number]
                        "item_image_button[2,3;1,1;default:dirt_with_grass;itemimagebutton;ItemImageButton]"..
                        "tooltip[0,11;3,2;Buz;#f00;#000]"..
                        "box[0,11;3,2;#00ff00]"..
+                       "hypertext[1,1;3,3;;" .. hypertext .. "]" ..^M
                        "container[0,18]"..
                                "box[1,2;3,2;#0a0a]"..
                                "scroll_container[1,2;3,2;scrbar2;horizontal;0.06]"..
```

Desour commented:
Fixed hypertext element not being clipped, as reported above. Scrolling hypertext also works now in scroll containers (the mousewheel event is handled, hence return true). Also added the test hypertext formspec provided above into minimal test formspec.

Desour commented on doc/lua_api.txt:
```diff
@@ -2101,6 +2101,26 @@ Elements
+  * be additionally moved by the current value of the scrollbar with the name
+    `scrollbar name` times `scroll factor` along the orientation `orientation` and
+  * be clipped to the rectangle defined by `X`, `Y`, `W` and `H`.
+* Possible values for `orientation` are `vertical` and `horizontal`.
```
Done.

Wuzzy2 commented:
I see there have been changes to `minimal` …

May I suggest for the future to keep all PRs that touch `minimal` on hold until DevTest (#9450) is merged? It's not fun when the code bases rapidly diverge …
